File: src/lib/objects.ts
// SOT: object-registry, object-kind-labels, object-kind-icons, object-sections, tool-registry, family-profile-lookup, admin-kinds
import type { Engine, FamilyProfile, ObjectKind, Tool } from "./bindings";
import { ENGINE_FACTS } from "./bindings/EngineFacts.gen";
import { FAMILY_PROFILES, SCOPED_OBJECT_KINDS } from "./bindings/FamilyProfiles.gen";
import type { IconName } from "./icons";

// WHAT:  UI metadata per object kind, keyed by the Rust `ObjectKind` enum.
// WHY:   `satisfies Record<ObjectKind, …>` fails the build when Rust gains a
//        kind this file does not label — the registry pattern, not a lookup
//        with a fallback.
// WHERE: src-tauri/src/model/objects.rs (ObjectKind), FamilyProfiles.gen.ts
export type ObjectSection = "containers" | "structure" | "code" | "security" | "server";

export interface ObjectKindMeta {
  label: string;
  plural: string;
  icon: IconName;
  section: ObjectSection;
}

const k = (label: string, plural: string, icon: IconName, section: ObjectSection): ObjectKindMeta => ({ label, plural, icon, section });

export const OBJECT_KINDS = {
  // ---- containers
  database: k("Database", "Databases", "database", "containers"),
  schema: k("Schema", "Schemas", "folder", "containers"),
  keyspace: k("Keyspace", "Keyspaces", "folder", "containers"),
  namespace: k("Namespace", "Namespaces", "folder", "containers"),
  bucket: k("Bucket", "Buckets", "archive", "containers"),
  dataset: k("Dataset", "Datasets", "layers", "containers"),
  column_family: k("Column family", "Column families", "columns", "containers"),
  // ---- structure
  table: k("Table", "Tables", "table", "structure"),
  view: k("View", "Views", "view", "structure"),
  materialized_view: k("Materialized view", "Materialized views", "view", "structure"),
  foreign_table: k("Foreign table", "Foreign tables", "globe", "structure"),
  virtual_table: k("Virtual table", "Virtual tables", "table", "structure"),
  partition: k("Partition", "Partitions", "layers", "structure"),
  collection: k("Collection", "Collections", "braces", "structure"),
  edge_collection: k("Edge collection", "Edge collections", "link", "structure"),
  graph: k("Graph", "Graphs", "hierarchy", "structure"),
  label: k("Label", "Labels", "tag", "structure"),
  relationship_type: k("Relationship type", "Relationship types", "link", "structure"),
  class: k("Class", "Classes", "cube", "structure"),
  measurement: k("Measurement", "Measurements", "chart", "structure"),
  metric: k("Metric", "Metrics", "activity", "structure"),
  topic: k("Topic", "Topics", "rows", "structure"),
  stream: k("Stream", "Streams", "flow", "structure"),
  channel: k("Channel", "Channels", "rss", "structure"),
  consumer_group: k("Consumer group", "Consumer groups", "users", "structure"),
  document: k("Document", "Documents", "file", "structure"),
  index: k("Index", "Indexes", "list", "structure"),
  constraint: k("Constraint", "Constraints", "shield", "structure"),
  sequence: k("Sequence", "Sequences", "hash", "structure"),
  type: k("Type", "Types", "swatch", "structure"),
  dictionary: k("Dictionary", "Dictionaries", "book", "structure"),
  projection: k("Projection", "Projections", "view", "structure"),
  alias: k("Alias", "Aliases", "tag", "structure"),
  snapshot: k("Snapshot", "Snapshots", "camera", "structure"),
  prefix: k("Prefix", "Prefixes", "text", "structure"),
  // ---- code
  function: k("Function", "Functions", "function", "code"),
  procedure: k("Procedure", "Procedures", "code", "code"),
  aggregate: k("Aggregate", "Aggregates", "function", "code"),
  trigger: k("Trigger", "Triggers", "flash", "code"),
  rule: k("Rule", "Rules", "route", "code"),
  event: k("Event", "Events", "bell", "code"),
  macro: k("Macro", "Macros", "code", "code"),
  package: k("Package", "Packages", "package", "code"),
  script: k("Script", "Scripts", "terminal", "code"),
  task: k("Task", "Tasks", "task", "code"),
  job: k("Job", "Jobs", "task", "code"),
  pipeline: k("Pipeline", "Pipelines", "flow", "code"),
  analyzer: k("Analyzer", "Analyzers", "text", "code"),
  synonym: k("Synonym", "Synonyms", "text", "code"),
  template: k("Template", "Templates", "file", "code"),
  recording_rule: k("Recording rule", "Recording rules", "chart", "code"),
  alert_rule: k("Alert rule", "Alert rules", "alert", "code"),
  // ---- security
  user: k("User", "Users", "user", "security"),
  role: k("Role", "Roles", "users", "security"),
  grant: k("Grant", "Grants", "key", "security"),
  policy: k("Policy", "Policies", "shield", "security"),
  api_key: k("API key", "API keys", "lock-key", "security"),
  acl: k("ACL", "ACLs", "lock", "security"),
  // ---- server
  session: k("Session", "Sessions", "activity", "server"),
  lock: k("Lock", "Locks", "lock", "server"),
  transaction: k("Transaction", "Transactions", "repeat", "server"),
  replica: k("Replica", "Replication", "database-sync", "server"),
  node: k("Node", "Nodes", "server", "server"),
  shard: k("Shard", "Shards", "layers", "server"),
  setting: k("Setting", "Settings", "settings", "server"),
  extension: k("Extension", "Extensions", "package", "server"),
  tablespace: k("Tablespace", "Tablespaces", "archive", "server"),
  publication: k("Publication", "Publications", "send", "server"),
  subscription: k("Subscription", "Subscriptions", "rss", "server"),
  replication_slot: k("Replication slot", "Replication slots", "database-sync", "server"),
  foreign_server: k("Foreign server", "Foreign servers", "globe", "server"),
  foreign_data_wrapper: k("Foreign data wrapper", "Foreign data wrappers", "plug", "server"),
  warehouse: k("Warehouse", "Warehouses", "server", "server"),
  stage: k("Stage", "Stages", "archive", "server"),
  quota: k("Quota", "Quotas", "gauge", "server"),
  slow_query: k("Slow query", "Slow queries", "clock", "server"),
  target: k("Target", "Targets", "target", "server"),
  alert: k("Alert", "Alerts", "alert", "server"),
  service: k("Service", "Services", "cube", "server"),
  backup: k("Backup", "Backups", "archive", "server"),
} satisfies Record<ObjectKind, ObjectKindMeta>;

const KIND_REGISTRY: Record<ObjectKind, ObjectKindMeta> = OBJECT_KINDS;

export const SECTIONS: readonly { id: ObjectSection; label: string }[] = [
  { id: "containers", label: "Containers" },
  { id: "structure", label: "Structure" },
  { id: "code", label: "Code" },
  { id: "security", label: "Security" },
  { id: "server", label: "Server" },
];

export function kindMeta(kind: ObjectKind): ObjectKindMeta {
  return KIND_REGISTRY[kind];
}

export function isScopedKind(kind: ObjectKind): boolean {
  return SCOPED_OBJECT_KINDS.includes(kind);
}

// WHAT:  Kinds the admin page owns (server-wide state, security, monitoring);
//        the object explorer sidebar shows the rest.
export const ADMIN_KINDS: readonly ObjectKind[] = [
  "session",
  "lock",
  "transaction",
  "slow_query",
  "replica",
  "node",
  "shard",
  "database",
  "user",
  "role",
  "grant",
  "acl",
  "api_key",
  "setting",
  "quota",
  "tablespace",
  "warehouse",
  "backup",
  "snapshot",
  "task",
  "job",
  "target",
  "alert",
  "service",
];

export function isAdminKind(kind: ObjectKind): boolean {
  return ADMIN_KINDS.includes(kind);
}

// WHAT:  Playground tabs, keyed by the Rust `Tool` enum.
export interface ToolMeta {
  label: string;
  icon: IconName;
  hint: string;
}

export const TOOLS = {
  stats: { label: "Server overview", icon: "activity", hint: "Connections, memory, cache, replication" },
  key_browser: { label: "Key browser", icon: "hash", hint: "Keys with type, TTL and a per-type editor" },
  erd: { label: "ER diagram", icon: "view", hint: "Tables and foreign keys, drawn" },
  vector_search: { label: "Vector search", icon: "radar", hint: "Similarity search with a payload filter" },
  search_playground: { label: "Search playground", icon: "search-list", hint: "Full-text query, filters, facets, highlights" },
  metrics_explorer: { label: "Metrics explorer", icon: "chart", hint: "Range queries charted over a time window" },
  message_viewer: { label: "Message viewer", icon: "message", hint: "Browse and produce topic messages" },
  pipeline_builder: { label: "Pipeline builder", icon: "flow", hint: "Aggregation stages, run and preview" },
  ledger_history: { label: "Ledger history", icon: "git-branch", hint: "Immutable history with transaction ids" },
  graph_view: { label: "Graph view", icon: "chart-relationship", hint: "Nodes and relationships from a query" },
  pub_sub: { label: "Pub/Sub", icon: "rss", hint: "Channels, publish, subscriber counts" },
  xml_viewer: { label: "XML viewer", icon: "xml", hint: "Documents as a collapsible tree" },
} satisfies Record<Tool, ToolMeta>;

const TOOL_REGISTRY: Record<Tool, ToolMeta> = TOOLS;

export function toolMeta(tool: Tool): ToolMeta {
  return TOOL_REGISTRY[tool];
}

export const TOOL_ORDER: readonly Tool[] = [
  "stats",
  "key_browser",
  "erd",
  "vector_search",
  "search_playground",
  "metrics_explorer",
  "message_viewer",
  "pipeline_builder",
  "ledger_history",
  "graph_view",
  "pub_sub",
  "xml_viewer",
];

// WHAT:  The static profile of an engine's adapter family: what the explorer
//        can list and which tools apply. Available before connecting.
// WHERE: src-tauri/src/integrations/<family>.rs (`profile()`)
export function profileOf(engine: Engine): FamilyProfile {
  return FAMILY_PROFILES[ENGINE_FACTS[engine].family];
}

export function objectKindsOf(engine: Engine): readonly ObjectKind[] {
  return profileOf(engine).objectKinds;
}

export function toolsOf(engine: Engine): readonly Tool[] {
  return profileOf(engine).tools;
}

export function hasTool(engine: Engine, tool: Tool): boolean {
  return toolsOf(engine).includes(tool);
}

export function hasKind(engine: Engine, kind: ObjectKind): boolean {
  return objectKindsOf(engine).includes(kind);
}

File: src-tauri/src/model/objects.rs
// SOT: object-kind, object-ref, object-summary, object-detail, object-action, code-language, server-stats, engine-tool, vector-search-request, search-request, range-query, ledger-history-request

use crate::model::{ColumnInfo, ResultSet};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use ts_rs::TS;

// ============================================================================
// OBJECT REGISTRY
//
// WHAT:  Every first-class thing a database exposes beyond rows: schemas,
//        views, functions, triggers, indexes, roles, sessions, topics, streams,
//        labels, buckets, metrics, snapshots… One enum for every family.
// WHY:   The object explorer, the admin page and the capability matrix are
//        generic: an adapter declares which kinds it has (`FamilyProfile`) and
//        answers `objects(kind, parent)` / `object_detail(ref)`. Adding a kind
//        here fails `src/lib/objects.ts` (`satisfies Record<ObjectKind, …>`)
//        until the UI has a label and an icon for it.
// HOW:   Kinds are engine-neutral; the engine-specific meaning lives in the
//        summary `badge` / detail `properties` (a Postgres `Type` is an enum or
//        a domain, a Snowflake `Stream` is a CDC stream, a Redis `Stream` is
//        the data type).
// WHERE: src-tauri/src/integrations/mod.rs (FamilyProfile, Integration::objects)
// ============================================================================
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export)]
pub enum ObjectKind {
    // ---- containers
    Database,
    Schema,
    Keyspace,
    Namespace,
    Bucket,
    Dataset,
    ColumnFamily,
    // ---- storage structures
    Table,
    View,
    MaterializedView,
    ForeignTable,
    VirtualTable,
    Partition,
    Collection,
    EdgeCollection,
    Graph,
    Label,
    RelationshipType,
    Class,
    Measurement,
    Metric,
    Topic,
    Stream,
    Channel,
    ConsumerGroup,
    Document,
    Index,
    Constraint,
    /// One column / property / attribute of a table, collection or class.
    Field,
    Sequence,
    Type,
    Dictionary,
    Projection,
    Alias,
    Snapshot,
    Prefix,
    // ---- code
    Function,
    Procedure,
    Aggregate,
    Trigger,
    Rule,
    Event,
    Macro,
    Package,
    Script,
    Task,
    Job,
    Pipeline,
    Analyzer,
    Synonym,
    Template,
    RecordingRule,
    AlertRule,
    // ---- security
    User,
    Role,
    Grant,
    Policy,
    ApiKey,
    Acl,
    // ---- server
    Session,
    Lock,
    Transaction,
    Replica,
    Node,
    Shard,
    Setting,
    Extension,
    Tablespace,
    Publication,
    Subscription,
    ReplicationSlot,
    ForeignServer,
    ForeignDataWrapper,
    Warehouse,
    Stage,
    Quota,
    SlowQuery,
    Target,
    Alert,
    Service,
    Backup,
}

impl ObjectKind {
    pub const ALL: [ObjectKind; 81] = [
        ObjectKind::Database,
        ObjectKind::Schema,
        ObjectKind::Keyspace,
        ObjectKind::Namespace,
        ObjectKind::Bucket,
        ObjectKind::Dataset,
        ObjectKind::ColumnFamily,
        ObjectKind::Table,
        ObjectKind::View,
        ObjectKind::MaterializedView,
        ObjectKind::ForeignTable,
        ObjectKind::VirtualTable,
        ObjectKind::Partition,
        ObjectKind::Collection,
        ObjectKind::EdgeCollection,
        ObjectKind::Graph,
        ObjectKind::Label,
        ObjectKind::RelationshipType,
        ObjectKind::Class,
        ObjectKind::Measurement,
        ObjectKind::Metric,
        ObjectKind::Topic,
        ObjectKind::Stream,
        ObjectKind::Channel,
        ObjectKind::ConsumerGroup,
        ObjectKind::Document,
        ObjectKind::Index,
        ObjectKind::Constraint,
        ObjectKind::Field,
        ObjectKind::Sequence,
        ObjectKind::Type,
        ObjectKind::Dictionary,
        ObjectKind::Projection,
        ObjectKind::Alias,
        ObjectKind::Snapshot,
        ObjectKind::Prefix,
        ObjectKind::Function,
        ObjectKind::Procedure,
        ObjectKind::Aggregate,
        ObjectKind::Trigger,
        ObjectKind::Rule,
        ObjectKind::Event,
        ObjectKind::Macro,
        ObjectKind::Package,
        ObjectKind::Script,
        ObjectKind::Task,
        ObjectKind::Job,
        ObjectKind::Pipeline,
        ObjectKind::Analyzer,
        ObjectKind::Synonym,
        ObjectKind::Template,
        ObjectKind::RecordingRule,
        ObjectKind::AlertRule,
        ObjectKind::User,
        ObjectKind::Role,
        ObjectKind::Grant,
        ObjectKind::Policy,
        ObjectKind::ApiKey,
        ObjectKind::Acl,
        ObjectKind::Session,
        ObjectKind::Lock,
        ObjectKind::Transaction,
        ObjectKind::Replica,
        ObjectKind::Node,
        ObjectKind::Shard,
        ObjectKind::Setting,
        ObjectKind::Extension,
        ObjectKind::Tablespace,
        ObjectKind::Publication,
        ObjectKind::Subscription,
        ObjectKind::ReplicationSlot,
        ObjectKind::ForeignServer,
        ObjectKind::ForeignDataWrapper,
        ObjectKind::Warehouse,
        ObjectKind::Stage,
        ObjectKind::Quota,
        ObjectKind::SlowQuery,
        ObjectKind::Target,
        ObjectKind::Alert,
        ObjectKind::Service,
        ObjectKind::Backup,
    ];

    // WHAT:  Kinds listed per namespace (the sidebar asks for them with the
    //        current schema / keyspace / database as `parent`). Everything else
    //        is listed once for the whole server.
    pub fn scoped(self) -> bool {
        matches!(
            self,
            ObjectKind::Table
                | ObjectKind::View
                | ObjectKind::MaterializedView
                | ObjectKind::ForeignTable
                | ObjectKind::VirtualTable
                | ObjectKind::Partition
                | ObjectKind::Collection
                | ObjectKind::EdgeCollection
                | ObjectKind::Index
                | ObjectKind::Constraint
                | ObjectKind::Field
                | ObjectKind::Sequence
                | ObjectKind::Type
                | ObjectKind::Function
                | ObjectKind::Procedure
                | ObjectKind::Aggregate
                | ObjectKind::Trigger
                | ObjectKind::Rule
                | ObjectKind::Event
                | ObjectKind::Macro
                | ObjectKind::Package
                | ObjectKind::Policy
                | ObjectKind::Measurement
                | ObjectKind::Document
        )
    }
}

// WHAT:  Playground-style tools that need their own tab, beyond the generic
//        grid / query / object views. Declared per family in `FamilyProfile`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export)]
pub enum Tool {
    /// Server overview: `Integration::server_stats` groups, auto-refreshing.
    Stats,
    /// Existing per-key inspector for key-value engines.
    KeyBrowser,
    /// Existing auto ER diagram from foreign keys.
    Erd,
    /// Similarity search: vector + top-k + payload filter → scored hits.
    VectorSearch,
    /// Full-text search: query + filters + facets + highlights.
    SearchPlayground,
    /// Range query over a time window with a step → chart.
    MetricsExplorer,
    /// Topic / partition / offset message browser with a producer.
    MessageViewer,
    /// Stage-by-stage aggregation pipeline editor.
    PipelineBuilder,
    /// Immutable history of a key / row with transaction ids and proofs.
    LedgerHistory,
    /// Node / relationship rendering of query results.
    GraphView,
    /// Publish to channels and list active subscriptions.
    PubSub,
    /// Tree view of XML documents.
    XmlViewer,
}

impl Tool {
    pub const ALL: [Tool; 12] = [
        Tool::Stats,
        Tool::KeyBrowser,
        Tool::Erd,
        Tool::VectorSearch,
        Tool::SearchPlayground,
        Tool::MetricsExplorer,
        Tool::MessageViewer,
        Tool::PipelineBuilder,
        Tool::LedgerHistory,
        Tool::GraphView,
        Tool::PubSub,
        Tool::XmlViewer,
    ];
}

// WHAT:  Identifies one object. `parent` is the namespace for scoped kinds
//        (schema, keyspace, database, bucket) or the owning object for nested
//        ones (the table of an index, the topic of a consumer group).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ObjectRef {
    pub kind: ObjectKind,
    pub name: String,
    pub parent: Option<String>,
}

// WHAT:  One row of an object list: enough for the sidebar and admin tables.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ObjectSummary {
    pub reference: ObjectRef,
    /// Short secondary text: signature, owner, row count, state, TTL…
    pub detail: Option<String>,
    /// Engine-specific subtype shown as a chip: `enum`, `btree`, `gsi`, `idle`, `leader`…
    pub badge: Option<String>,
}

impl ObjectSummary {
    pub fn new(kind: ObjectKind, name: impl Into<String>, parent: Option<String>) -> ObjectSummary {
        ObjectSummary { reference: ObjectRef { kind, name: name.into(), parent }, detail: None, badge: None }
    }
    pub fn with_detail(mut self, detail: impl Into<String>) -> ObjectSummary {
        self.detail = Some(detail.into());
        self
    }
    pub fn with_badge(mut self, badge: impl Into<String>) -> ObjectSummary {
        self.badge = Some(badge.into());
        self
    }
}

// WHAT:  Syntax the definition text is in (drives highlighting in the UI).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export)]
pub enum CodeLanguage {
    Sql,
    Json,
    Xml,
    Text,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ObjectProperty {
    pub name: String,
    pub value: String,
}

// WHAT:  Something the user can do to the object. `statement` is in the
//        engine's native command language and runs through `execute_query`,
//        so the read-only lock, the destructive confirmation and the history
//        log all apply without the adapter doing anything.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ObjectAction {
    pub id: String,
    pub label: String,
    pub destructive: bool,
    pub statement: String,
}

impl ObjectAction {
    pub fn new(id: &str, label: &str, statement: impl Into<String>) -> ObjectAction {
        ObjectAction { id: id.to_string(), label: label.to_string(), destructive: false, statement: statement.into() }
    }
    pub fn destructive(id: &str, label: &str, statement: impl Into<String>) -> ObjectAction {
        ObjectAction { id: id.to_string(), label: label.to_string(), destructive: true, statement: statement.into() }
    }
}

// WHAT:  Everything the object tab shows: definition source, a property
//        sheet, an optional column list (table-like kinds), an optional
//        tabular payload (partitions, members, entries, mapping fields),
//        nested objects (a table's indexes, a topic's consumer groups) and
//        the actions available.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ObjectDetail {
    pub reference: ObjectRef,
    pub definition: Option<String>,
    pub language: CodeLanguage,
    pub properties: Vec<ObjectProperty>,
    pub columns: Vec<ColumnInfo>,
    pub rows: Option<ResultSet>,
    pub children: Vec<ObjectSummary>,
    pub actions: Vec<ObjectAction>,
}

impl ObjectDetail {
    pub fn empty(reference: &ObjectRef) -> ObjectDetail {
        ObjectDetail {
            reference: reference.clone(),
            definition: None,
            language: CodeLanguage::Text,
            properties: Vec::new(),
            columns: Vec::new(),
            rows: None,
            children: Vec::new(),
            actions: Vec::new(),
        }
    }
    pub fn definition(mut self, text: impl Into<String>, language: CodeLanguage) -> ObjectDetail {
        self.definition = Some(text.into());
        self.language = language;
        self
    }
    pub fn property(mut self, name: &str, value: impl Into<String>) -> ObjectDetail {
        self.properties.push(ObjectProperty { name: name.to_string(), value: value.into() });
        self
    }
    pub fn action(mut self, action: ObjectAction) -> ObjectDetail {
        self.actions.push(action);
        self
    }
}

// WHAT:  One monitoring figure. `numeric` lets the UI draw a sparkline across
//        refreshes; `value` is the display text (already formatted by the adapter).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct Stat {
    pub label: String,
    pub value: String,
    pub unit: Option<String>,
    pub hint: Option<String>,
    pub numeric: Option<f64>,
}

impl Stat {
    pub fn text(label: &str, value: impl Into<String>) -> Stat {
        Stat { label: label.to_string(), value: value.into(), unit: None, hint: None, numeric: None }
    }
    pub fn number(label: &str, value: f64, unit: Option<&str>) -> Stat {
        Stat { label: label.to_string(), value: format_number(value), unit: unit.map(str::to_string), hint: None, numeric: Some(value) }
    }
    pub fn with_hint(mut self, hint: impl Into<String>) -> Stat {
        self.hint = Some(hint.into());
        self
    }
}

// WHAT:  Grouped server statistics (Connections / Memory / Cache / Replication …).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct StatGroup {
    pub title: String,
    pub stats: Vec<Stat>,
}

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ServerStats {
    pub groups: Vec<StatGroup>,
    /// ISO-8601 timestamp the figures were read at.
    pub collected_at: String,
}

impl ServerStats {
    pub fn now(groups: Vec<StatGroup>) -> ServerStats {
        ServerStats { groups, collected_at: chrono::Utc::now().to_rfc3339() }
    }
}

// WHAT:  Plain `1234567.5` → `1,234,567.5`; integers lose the fraction.
pub fn format_number(value: f64) -> String {
    let whole = value.trunc().abs() as u64;
    let mut digits = whole.to_string();
    let mut grouped = String::with_capacity(digits.len() + digits.len() / 3);
    while digits.len() > 3 {
        let tail = digits.split_off(digits.len() - 3);
        grouped = if grouped.is_empty() { tail } else { format!("{tail},{grouped}") };
    }
    grouped = if grouped.is_empty() { digits } else { format!("{digits},{grouped}") };
    let sign = if value < 0.0 { "-" } else { "" };
    let fraction = value.fract().abs();
    if fraction < f64::EPSILON {
        format!("{sign}{grouped}")
    } else {
        format!("{sign}{grouped}.{}", format!("{fraction:.2}").trim_start_matches("0.").trim_end_matches('0'))
    }
}

// ---- playground requests ----------------------------------------------------

// WHAT:  Similarity search against one collection. `vector` is the query
//        embedding; `filter` is the engine's native payload filter (JSON).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct VectorSearchRequest {
    pub collection: String,
    pub vector: Vec<f64>,
    /// Named vector / field to search (engines with several per point).
    pub vector_name: Option<String>,
    pub top_k: u32,
    pub filter: Option<JsonValue>,
    pub include_vectors: bool,
}

// WHAT:  Full-text search against one index: free text plus the engine's
//        native filter syntax, optional facets and highlighting.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct SearchRequest {
    pub index: String,
    pub query: String,
    pub filter: Option<String>,
    pub facets: Vec<String>,
    pub sort: Vec<String>,
    pub highlight: bool,
    pub limit: u32,
    pub offset: u32,
}

// WHAT:  Hits plus per-facet counts, so the playground can render both.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct SearchResult {
    pub hits: ResultSet,
    pub total: Option<u64>,
    pub facets: Vec<FacetCounts>,
    pub took_ms: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct FacetCounts {
    pub field: String,
    pub values: Vec<FacetValue>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct FacetValue {
    pub value: String,
    pub count: u64,
}

// WHAT:  Range query for metrics engines: expression over [start, end] with a
//        step, all in seconds since the epoch.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct RangeQueryRequest {
    pub query: String,
    pub start: f64,
    pub end: f64,
    pub step_seconds: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct Series {
    /// Label set / tag set identifying the series, as `key=value` text.
    pub name: String,
    pub labels: Vec<ObjectProperty>,
    /// `[timestamp_seconds, value]` pairs, ascending.
    pub points: Vec<[f64; 2]>,
}

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct RangeResult {
    pub series: Vec<Series>,
    pub warnings: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn all_lists_every_kind_once() {
        let mut seen = std::collections::HashSet::new();
        for kind in ObjectKind::ALL {
            assert!(seen.insert(kind), "{kind:?} listed twice");
        }
        assert_eq!(seen.len(), ObjectKind::ALL.len());
    }

    #[test]
    fn numbers_group_thousands() {
        assert_eq!(format_number(0.0), "0");
        assert_eq!(format_number(999.0), "999");
        assert_eq!(format_number(1234567.0), "1,234,567");
        assert_eq!(format_number(1234.5), "1,234.5");
        assert_eq!(format_number(-42.25), "-42.25");
    }

    #[test]
    fn detail_builders_accumulate() {
        let r = ObjectRef { kind: ObjectKind::Table, name: "t".into(), parent: Some("public".into()) };
        let d = ObjectDetail::empty(&r)
            .definition("CREATE TABLE t ()", CodeLanguage::Sql)
            .property("owner", "postgres")
            .action(ObjectAction::destructive("drop", "Drop", "DROP TABLE t"));
        assert_eq!(d.language, CodeLanguage::Sql);
        assert_eq!(d.properties.len(), 1);
        assert!(d.actions[0].destructive);
    }
}
